### PR TITLE
Seeds menu

### PR DESCRIPTION
**Pull request by LuKks**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -56,24 +64,50 @@ async function start () {
...
   }
 
-  swarm.on('connection', onsocket)
-  swarm.listen()
+  for (const key of lists) {
+    const bee = await downloadLists(key)
+
+    if (argv['dry-run']) continue
+
+    for (const type of ['core', 'bee', 'drive', 'seeder']) {
```
Comment (mafintosh) [inline index.js]: instead of using a sub for each, can we have the type in the value instead? then you just need to know the key for interact with it

Comment (mafintosh) [reply index.js]: and then have no subs

File: index.js [hunk 2]
```diff
@@ -53,170 +43,75 @@ async function main () {
...
-
-      if (!resource) {
-
-      }
-
-      if (resource) {
-        // TODO: skip same
-      }
+async function update (tracker, list) {
+  for (const [, info] of tracker.resources) {
```
Comment (mafintosh) [inline index.js]: expose a .values() helper for this that does tracker.resources.values()

File: lib/tracker.js [hunk 3]
```diff
@@ -1,35 +1,172 @@
...
+      info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    }
 
-    this.resources.push(info)
+    this.resources.set(id, info)
 
     return info
   }
 
+  async remove (key) {
```
Comment (mafintosh) [inline lib/tracker.js]: remove needs to aquire the lock also i think

File: lib/tracker.js [hunk 4]
```diff
@@ -1,35 +1,172 @@
...
+    const info = this.get(key)
+
+    if (info.seeders) await info.seeders.destroy()
+    await info.instance.close()
+
+    this.resources.delete(info.key)
```
Comment (mafintosh) [inline lib/tracker.js]: resources might have been overwritten by the above awaits. check if the resource is what you expect it to be first, `if (this.resources.get(info.key) === info) this.resources.delete(info.key)`

File: lib/tracker.js [hunk 5]
```diff
@@ -1,35 +1,172 @@
...
+  }
+
+  get (key) {
+    const id = HypercoreId.normalize(key)
+    const info = this.resources.get(id)
+    if (!info) throw new Error('Resource does not exists')
+    return info
+  }
+
   getByType (type) {
```
Comment (mafintosh) [inline lib/tracker.js]: listByType?

File: lib/tracker.js [hunk 6]
```diff
@@ -1,35 +1,172 @@
...
+
+  _suitCore (core, info, opts) {
+    followSpeed(core, info)
+    if (opts.announce !== false) this._swarmCore(core)
+    core.download()
+  }
+
+  _swarmCore (core) {
+    const done = core.findingPeers()
+    this.swarm.join(core.discoveryKey, { client: true, server: !this.backup })
```
Comment (mafintosh) [inline lib/tracker.js]: you need to store the discovery session returned by join and destroy it on if the resource is removed

File: lib/tracker.js [hunk 7]
```diff
@@ -1,35 +1,172 @@
...
+    }
+
+    return drive
   }
+
+  async _createSeeders (publicKey) {
+    const id = HypercoreId.encode(publicKey)
+    const keyPair = await this.store.createKeyPair('simple-seeder-swarm@' + id)
+
+    const sw = new Seeders(publicKey, { dht: this.swarm.dht, keyPair })
```
Comment (mafintosh) [inline lib/tracker.js]: are you destroying the seeder swarm on resource deletion?

Comment (LuKks) [reply lib/tracker.js]: Yes, on remove method there was this:
https://github.com/holepunchto/simple-seeder/blob/144f59bdb3a07c903eea6a08baa378df422a7c65/lib/simple-seeder.js#L66

File: index.js (hunk 2 above)
Comment (LuKks) [reply index.js]: Instead of the helper just using `tracker.resources.values()` directly now, or you actually prefer the `values` method?

File: lib/tracker.js (hunk 4 above)
Comment (LuKks) [reply lib/tracker.js]: I added the lock on remove, I think it's worth it as it keeps the code without extra checks

File: lib/tracker.js (hunk 5 above)
Comment (LuKks) [reply lib/tracker.js]: Yea, I removed `getByType` and `find`

Now it's just one `filter` method (like Array filter)

File: index.js (hunk 2 above)
Comment (mafintosh) [reply index.js]: would say that resources is internal and an outsider shouldn't iterate it

File: index.js [hunk 8]
```diff
@@ -1,168 +1,163 @@
...
+  const snap = list.snapshot()
 
-    if (track) {
-      const info = { core, blocks: { download: speedometer(), upload: speedometer() }, network: { download: speedometer(), upload: speedometer() } }
-      core.on('download', onspeed.bind(null, 'download', info))
-      core.on('upload', onspeed.bind(null, 'upload', info))
-      tracking.cores.push(info)
-    }
+  for (const info of tracker.values()) {
+    if (info.external) continue // List does not control resources from argv or file, including itself
```
Comment (mafintosh) [inline index.js]: unsure what this means and what external is

File: index.js [hunk 9]
```diff
@@ -1,168 +1,163 @@
...
 
-    if (announce !== false) swarm.join(core.discoveryKey, { client: true, server: !argv.backup })
-    core.download()
+    const e = await snap.get(info.key)
+
+    if (e === null || e.value.type !== info.type) {
```
Comment (mafintosh) [inline index.js]: feels like the type mismatch should be handled by re-adding it

File: index.js [hunk 10]
```diff
@@ -1,168 +1,163 @@
...
   }
 
-  async function downloadCore (core, announce, { track = true } = {}) {
-    core = typeof core === 'string' ? store.get(HypercoreId.decode(core)) : core
+  return seeds
+}
 
-    await core.ready()
+async function update (tracker, list) {
+  const snap = list.snapshot()
```
Comment (mafintosh) [inline index.js:121]: when is this closed?

File: index.js [hunk 11]
```diff
@@ -1,168 +1,143 @@
...
-
   swarm.on('connection', onsocket)
   swarm.listen()
+  goodbye(() => swarm.destroy())
 
-  for (const key of cores) await downloadCore(key)
-  for (const key of bees) await downloadBee(key)
-  for (const key of drives) await downloadDrive(key)
+  if (argv.menu) {
+    const menu = require('./menu.js')
```
Comment (HDegroote) [inline index.js:50]: Any particular reason for requiring here instead of at the top? Seems strange to me

File: index.js [hunk 12]
```diff
@@ -1,168 +1,143 @@
...
 
-    if (!drives.includes(key)) {
-      await downloadDrive(key, false)
-    }
+  const intervalId = setInterval(ui, argv.i || 5000)
+  goodbye(() => clearInterval(intervalId)) // Explicit cleanup
+  ui()
+
+  const lists = tracker.filter(r => r.type === 'list')
+  if (lists[0] && !argv['dry-run']) {
```
Comment (HDegroote) [inline index.js]: More of a detail, and no need to change, but I prefer to extract the relevant argv's at the start of my code and to then pass them around as variables instead of accessing argv throughout the code, creates more of a separation between the input and the processing code

File: lib/simple-seeder.js [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    }
+    return arr
+  }
+
+  values () {
+    return this.resources.values()
+  }
+
+  async destroy () {
+    if (this.destroying) return
```
Comment (HDegroote) [inline lib/simple-seeder.js:142]: Shouldn't it store the promise at the end of this method and return that instead of just returning? (or could use ready resource to avoid having to think about stuff like that)

File: lib/simple-seeder.js [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+      release()
+    }
+  }
+
+  async _put (key, opts) {
+    if (!this.has(key)) return this._add(key, opts)
+
+    const info = this.get(key)
+
+    if (info.type !== opts.type) {
```
Comment (HDegroote) [inline lib/simple-seeder.js:79]: How does this case happen? Seems strange to have a key be added as one type, and then later added again as another type

File: lib/simple-seeder.js [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+      await this._remove(key)
+      return this._add(key, opts)
+    }
+
+    if (!!info.seeders !== opts.seeders) {
```
Comment (HDegroote) [inline lib/simple-seeder.js:84]: The if-else branching and the conditions here are hard to follow. Can this be flattened a bit more? For example, if you put `info.description = opts.description || null` first, you can just early return I think:

`    if (!!info.seeders === opts.seeders) return info`

 Also a bit confused by the underlying logic: when does it happen that info.seeders exists, and opts.seeders is not set, meaning we want to remove it?

File: lib/simple-seeder.js [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+  get (key) {
+    const id = HypercoreId.normalize(key)
+    const info = this.resources.get(id)
+    if (!info) throw new Error('Resource does not exists')
+    return info
+  }
+
+  filter (onfilter) {
+    const arr = []
+    for (const info of this.values()) {
```
Comment (HDegroote) [inline lib/simple-seeder.js]: I think you can just use .filter() here, like 
`return [...this.values()].filter(onfilter)

File: lib/simple-seeder.js [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    const keyPair = await this.store.createKeyPair('simple-seeder-swarm@' + id)
+
+    const sw = new Seeders(publicKey, { dht: this.swarm.dht, keyPair })
+    sw.join()
+    sw.on('connection', (socket) => this.store.replicate(socket))
+
+    return sw
+  }
+
+  _suitCore (core, info, opts) {
```
Comment (HDegroote) [inline lib/simple-seeder.js:208]: suit core sounds strange, if the concept is that of following a core, I would call it `followCore`.
The logic for this is also hard to follow, I can't figure out why `followSpeed` and `onSpeed` helper methods were created. From what I understand, it could be more explicit by just defining 2 handlers here:

```
  core.on('download', (...) => {   
    info.blocks['down'](1) })
    info.network['down'](byteLength)
  }
  core.on('upload', (...) => {
    info.blocks['up'](1) })
    info.network['up'](byteLength)
  })
```

**Review by HDegroote — COMMENTED** [review]
Didn't finish completely, but need to go now. Will come back on it later

File: index.js (hunk 11 above)
Comment (LuKks) [reply index.js:50]: `tiny-menu` uses `readline` which is not supported in some enviroments relevant for us

Comment (LuKks) [reply index.js:50]: So we load it only when `--menu` so you can still run SimpleSeeder anywhere

File: lib/simple-seeder.js (hunk 13 above)
Comment (LuKks) [reply lib/simple-seeder.js:142]: Sure but we don't gain anything from doing that on this case

File: lib/simple-seeder.js (hunk 14 above)
Comment (LuKks) [reply lib/simple-seeder.js:79]: The commonest scenario would probably be promoting a core to a bee (and vice versa):
`await tracker.put(key, { type: 'core' })`
And later
`await tracker.put(key, { type: 'bee' })`

I guess someone could add `drive.core` and `drive.blobs` as isolated bees

Unsure if there is any other scenario but the point is that it can happen so we can easily handle it

File: lib/simple-seeder.js (hunk 15 above)
Comment (LuKks) [reply lib/simple-seeder.js:84]: `await tracker.put(key, { type: 'drive', seeders: true })`
`await tracker.put(key, { type: 'drive', seeders: false })`

Comment (LuKks) [reply lib/simple-seeder.js:84]: `!!info.seeders !== opts.seeders` looks complicated I know but it doesn't get more complicated than that, it just checks that they're different in boolean values

(`info.seeders` is null or an object, and `opts.seeders` a bool)

Similar as the check above `info.type !== opts.type`

File: lib/simple-seeder.js (hunk 17 above)
Comment (LuKks) [reply lib/simple-seeder.js:208]: That still creates new functions and repeats the code twice, so I tend to prefer a bind reusing the same code

But no need for the `followSpeed` helper, will move it to the internal method

Comment (LuKks) [reply lib/simple-seeder.js:208]: I kind of think that Hypercore should do the `speedometer` internally so it just gives you the speeds

File: lib/simple-seeder.js (hunk 16 above)
Comment (LuKks) [reply lib/simple-seeder.js]: Functional programming is illegal you know

File: lib/simple-seeder.js (hunk 15 above)
Comment (LuKks) [reply lib/simple-seeder.js:84]: It basically toggles the seeders

File: index.js (hunk 11 above)
Comment (HDegroote) [reply index.js:50]: Aha, that makes a lot of sense. Thnx for clarifying! Could consider adding a comment to explain, but fine as is too

File: lib/simple-seeder.js (hunk 13 above)
Comment (HDegroote) [reply lib/simple-seeder.js:142]: Ah, I had misunderstood the intent of the `if 'this.destroying` logic. I thought it was to avoid calling destroy twice, but it's to avoid `_add` being run when already destroying. If the behaviour of calling destroy twice is undefined, I'm fine with the code as-is (The current contract of calling destroy twice is that you're not sure that when the promise resolves, the instance has in fact been destroyed)

Comment (HDegroote) [reply lib/simple-seeder.js:142]: Correction: the destroy code doesn't work, because `info` objects do not have a destroy method, so it crashes. The error gets swallowed though. I couldn't repro easily with the menu, but easy to repro by hosting a bee.

Change the goodbye definition on tracker to
```
  goodbye(async () => {
    console.log('Calling tracker destroy')
    try {
      await tracker.destroy()
    } catch (e) {
      console.error('caught error in tracker goodbye:')
      console.error(e)
    }
  })
```
Then run

`node index.js -b 7d2050aa131f1d0dad47723c3ac0a4af2dd087d4269848c9909c6b28c2463b34`

and `ctrl-c` it.

Expected output:

```^CCalling tracker destroy
destroying
caught error in tracker goodbye:
TypeError: info.destroy is not a function
    at SimpleSeeder.destroy (/home/hans/holepunch/simple-seeder/lib/simple-seeder.js:150:25)
    at Object.fn (/home/hans/holepunch/simple-seeder/index.js:62:21)
    at /home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:37:47
    at Array.map (<anonymous>)
    at loop (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:37:36)
    at onexit (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:33:3)
    at onsignal (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:17:3)
    at process.emit (node:events:513:28)
```

Comment (HDegroote) [reply lib/simple-seeder.js:142]: Fwiw, it doesn't crash when using menu because menu doesn't call tracker.destroy(), but I think that's expected since menu is just for managing your list

Comment (LuKks) [reply lib/simple-seeder.js:142]: Wow that's true, the `goodbye` was ignoring all errors (we should have added your PR lol, we should review that one)

If there is `info.instance` then it should do `info.instance.close || .destroy`

File: index.js [hunk 18]
```diff
@@ -75,6 +75,7 @@ async function main () {
     const list = lists[0].instance
     const debounced = debounceify(update.bind(null, tracker, list))
     list.core.on('append', debounced)
+    goodbye(() => list.core.off('append', debounced))
```
Comment (mafintosh) [inline index.js:78]: not needed

File: lib/simple-seeder.js (hunk 13 above)
Comment (mafintosh) [reply lib/simple-seeder.js:142]: i'm merging this now as i need it, feel free to fix the issue
